gobblin-api/src/main/java/gobblin/testing/AssertWithBackoff.java

```diff
@@ -0,0 +1,200 @@
+package gobblin.testing;
+
+import java.util.concurrent.TimeoutException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Function;
+import com.google.common.base.Optional;
+import com.google.common.base.Predicate;
+import com.google.common.base.Predicates;
+
+/**
+ * A helper class to perform a check for a condition with a back-off. Primary use of this is when a tests
+   * needs for something asynchronous to happen, say a service to start.
+ **/
+public class AssertWithBackoff {
+
+  /** the max time in milliseconds to wait for the condition to become true */
+  private long timemoutMs = 30 * 000;
+  /** a logger to use for logging waiting, results, etc. */
+  private Logger _log = LoggerFactory.getLogger(AssertWithBackoff.class);
```

Comment:
I know member names starting with `_` are LinkedIn's standard, but Gobblin has not been following this standard since the beginning, and our coding style guide does not follow it either.